zcash_client_sqlite/src/wallet.rs

```diff
@@ -157,29 +163,86 @@ pub(crate) fn add_account<P: consensus::Parameters>(
+    let chain_tip = scan_queue_extrema(conn)?.map(|(_, max)| max);
 
-pub(crate) fn add_account_internal<P: consensus::Parameters, E: From<rusqlite::Error>>(
-    conn: &rusqlite::Transaction,
-    network: &P,
-    accounts_table: &'static str,
-    account: AccountId,
-    key: &UnifiedFullViewingKey,
-) -> Result<(), E> {
-    let ufvk_str: String = key.encode(network);
     conn.execute(
-        &format!(
-            "INSERT INTO {} (account, ufvk) VALUES (:account, :ufvk)",
-            accounts_table
-        ),
-        named_params![":account": &<u32>::from(account), ":ufvk": &ufvk_str],
+        "INSERT INTO accounts (account, ufvk, birthday_height, recover_until_height)
+        VALUES (:account, :ufvk, :birthday_height, :recover_until_height)",
+        named_params![
+            ":account": u32::from(account),
+            ":ufvk": &key.encode(params),
+            ":birthday_height": u32::from(birthday.height()),
+            ":recover_until_height": birthday.recover_until().map(u32::from)
+        ],
     )?;
 
+    // If a birthday frontier is available, insert it into the note commitment tree. If the
+    // birthday frontier is the empty frontier, we don't need to do anything.
+    if let Some(frontier) = birthday.sapling_frontier().value() {
+        debug!("Inserting frontier into ShardTree: {:?}", frontier);
+        let shard_store = SqliteShardStore::<
+            _,
+            zcash_primitives::sapling::Node,
+            SAPLING_SHARD_HEIGHT,
+        >::from_connection(conn, SAPLING_TABLES_PREFIX)?;
+        let mut shard_tree: ShardTree<
+            _,
+            { zcash_primitives::sapling::NOTE_COMMITMENT_TREE_DEPTH },
+            SAPLING_SHARD_HEIGHT,
+        > = ShardTree::new(shard_store, PRUNING_DEPTH.try_into().unwrap());
+        shard_tree.insert_frontier_nodes(
+            frontier.clone(),
+            Retention::Checkpoint {
+                // This subtraction is safe, because the non-empty frontier cannot exist at genesis
+                id: birthday.height() - 1,
+                is_marked: false,
+            },
+        )?;
+    }
+
+    let sapling_activation_height = params
+        .activation_height(NetworkUpgrade::Sapling)
+        .expect("Sapling activation height must be available.");
+
+    // Add the ignored range up to and including the birthday height.
+    if sapling_activation_height < birthday.height() {
+        let ignored_range = sapling_activation_height..birthday.height();
+
+        replace_queue_entries::<SqliteClientError>(
+            conn,
+            &ignored_range,
+            Some(ScanRange::from_parts(
+                ignored_range.clone(),
+                ScanPriority::Ignored,
+            ))
+            .into_iter(),
+            false,
+        )?;
+    };
+
+    // Rewrite the scan ranges above the birthday height so that we'll ensure we re-scan to find
+    // any notes that might belong to the newly added account.
```

Comment:
```suggestion
    // Rewrite the scan ranges starting from the birthday height so that we'll ensure we
    // re-scan to find any notes that might belong to the newly added account.
```